File: #0019.remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {

        vector<ListNode *> v;
        ListNode *ln = head;
        while (ln) {
            v.push_back(ln);
            ln = ln->next;
        }

        if (n == v.size()) {
            return head->next;
        }
        ListNode* v_remove = v[v.size()-n];
        ListNode* v_parent = v[v.size() - n - 1];
        v_parent->next = v_remove->next;
        delete v_remove;
        return head;
    }
};

File: #0019.remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let cnt = 0;
    for (node = head; node != null; node = node.next) {
        cnt++;
    }

    if (cnt - n < 1) {
        return head.next;
    }

    for (node = head, i = 1; node != null; node = node.next, i++) {
        if (i == cnt - n) {
            node.next = node.next.next;
            break;
        }
    }
    return head;
};
